core/client/fs/src/main/java/alluxio/client/file/cache/store/MemoryPageStore.java

```diff
@@ -0,0 +1,132 @@
+
+import alluxio.client.file.cache.PageId;
+import alluxio.client.file.cache.PageInfo;
+import alluxio.client.file.cache.PageStore;
+import alluxio.exception.PageNotFoundException;
+import alluxio.exception.status.ResourceExhaustedException;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+/**
+ * The {@link MemoryPageStore} is an implementation of {@link PageStore} which
+ * stores all pages in memory.
+ */
+@NotThreadSafe
+public class MemoryPageStore implements PageStore {
+  private static final Logger LOG = LoggerFactory.getLogger(MemoryPageStore.class);
+  private final long mPageSize;
+  private final long mCapacity;
+  private final int mFileBuckets;
+
+  private ConcurrentHashMap<String, byte[]> mPageStoreMap = null;
+
+  /**
+   * Creates a new instance of {@link MemoryPageStore}.
+   *
+   * @param options options for the buffer store
+   */
+  public MemoryPageStore(MemoryPageStoreOptions options) {
+    mPageSize = options.getPageSize();
+    mCapacity = (long) (options.getCacheSize() / (1 + options.getOverheadRatio()));
+    mFileBuckets = options.getFileBuckets();
+    mPageStoreMap = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public void put(PageId pageId, byte[] page) throws ResourceExhaustedException, IOException {
+    String pageKey = getKeyFromPageId(pageId);
+    try {
+      byte[] mPage = new byte[page.length];
+      System.arraycopy(page, 0, mPage, 0, page.length);
+      mPageStoreMap.put(pageKey, mPage);
+    } catch (Exception e) {
+      throw new IOException("Failed to put cached data in memory for page " + pageId);
+    }
+  }
+
+  @Override
+  public int get(PageId pageId, int pageOffset, int bytesToRead, byte[] buffer, int bufferOffset)
+      throws IOException, PageNotFoundException {
+    Preconditions.checkArgument(buffer != null, "buffer is null");
+    Preconditions.checkArgument(pageOffset >= 0, "page offset should be non-negative");
+    Preconditions.checkArgument(buffer.length >= bufferOffset,
+        "page offset %s should be " + "less or equal than buffer length %s", bufferOffset,
+        buffer.length);
+    String pageKey = getKeyFromPageId(pageId);
+    if (!mPageStoreMap.containsKey(pageKey)) {
+      throw new PageNotFoundException(pageKey);
+    }
+    byte[] mPage = mPageStoreMap.get(pageKey);
+    Preconditions.checkArgument(pageOffset <= mPage.length, "page offset %s exceeded page size %s",
+        pageOffset, mPage.length);
+    int bytesLeft = (int) Math.min(mPage.length - pageOffset, buffer.length - bufferOffset);
+    bytesLeft = Math.min(bytesLeft, bytesToRead);
+    System.arraycopy(mPage, pageOffset, buffer, bufferOffset, bytesLeft);
```

Comment:
arraycopy might make it safer, but is it necessary?

Reply:
It's necessary, the reason is the same as the putting method, shown comments above.

Reply:
but you always make a copy in the put method, I don't think the byte array would be got modified unless there is some other client to modify it (I don't think presto will modify any bytes array it reads from)

Reply:
> I don't think the byte array would be got modified unless there is some other client to modify it (I don't think presto will modify any bytes array it reads from)

Yeah, like Spark, Presto, they will not modify any data, but we can not guarantee all of the systems will not modify any data after reading from, shown as below:
```
        System.out.println(Arrays.toString(mPageDataMap.get(1l)));
        System.out.println(Arrays.toString(mPageDataMap.get(2l)));

        // read data from memory cache data
        byte[] data1 = mPageDataMap.get(1l);
        // modify the data directly
        data1[0] = 2;
        byte[] data2 = mPageDataMap.get(2l);
        // modify the data directly
        data2[1] = 3;
        
        System.out.println(Arrays.toString(mPageDataMap.get(1l)));
        System.out.println(Arrays.toString(mPageDataMap.get(2l)));
```

the results are:

![image](https://user-images.githubusercontent.com/9430290/152938310-38c429bf-2acd-4f9c-b226-a092cb28ce2a.png)

Reply:
gotcha, let's keep the array copy for now,  we could improve it if we saw any GC or memory issues.  Also thinking to make the bytes immutable, but it's really difficult to make an immutable array in java.